### PR TITLE
[DATA-453] Get lazy png image for mono case

### PR DESCRIPTION
**Pull request by kkufieta**
https://viam.atlassian.net/browse/DATA-453

### CODE REVIEW

File: services/slam/builtin/builtin.go [hunk 1]
```diff
@@ -868,17 +874,22 @@ func (slamSvc *builtIn) getAndSaveDataDense(ctx context.Context, cams []camera.C
...
+	switch slamMode {
+	case slam.Rgbd:
 		colorFilename := filepath.Join(dataDirectory, "data", "rgb", cameraName+"_data_"+timeStamp.UTC().Format(slamTimeFormat)+fileType)
 		depthFilename := filepath.Join(dataDirectory, "data", "depth", cameraName+"_data_"+timeStamp.UTC().Format(slamTimeFormat)+fileType)
 		return []string{colorFilename, depthFilename}
+	case slam.Mono:
+		colorFilename := filepath.Join(dataDirectory, "data", "rgb", cameraName+"_data_"+timeStamp.UTC().Format(slamTimeFormat)+fileType)
+		return []string{colorFilename}
+	default:
+		filename := filepath.Join(dataDirectory, "data", cameraName+"_data_"+timeStamp.UTC().Format(slamTimeFormat)+fileType)
```
Comment (JohnN193) [inline services/slam/builtin/builtin.go:890]: [q] should the default be rgb? or is it better to leave this in data for when we add another server?

**Review by JohnN193 — APPROVED** [review]
LGTM, one quick question

File: services/slam/builtin/builtin.go [hunk 2]
```diff
@@ -202,6 +209,7 @@ func runtimeServiceValidation(
 			path, err = slamSvc.getAndSaveDataDense(ctx, cams)
 			paths = append(paths, path)
 		default:
+			// TODO[kat]: Doesn't this mean that the slam AlgoType is not defined?
```
Comment (tessavitabile) [inline services/slam/builtin/builtin.go:212]: [nit] We shouldn't leave TODOs in the code unless they're tied to a particular ticket.

File: services/slam/builtin/builtin.go [hunk 3]
```diff
@@ -686,6 +694,26 @@ func (slamSvc *builtIn) StopSLAMProcess() error {
...
 
+func (slamSvc *builtIn) getLazyPngImage(ctx context.Context, cam camera.Camera) ([]byte, error) {
+	// We will hint that we want a PNG.
+	// The Camera service server implementation in RDK respects this; others may not.
+	img, release, err := camera.ReadImage(
+		gostream.WithMIMETypeHint(ctx, utils.WithLazyMIMEType(utils.MimeTypePNG)), cam)
+	if err != nil {
+		return nil, err
+	}
+	defer release()
```
Comment (tessavitabile) [inline services/slam/builtin/builtin.go:705]: I'm concerned about calling `release()` before we've written out the image to a file. Maybe `getLazyPngImage()` should return `release` instead.

File: services/slam/builtin/builtin.go [hunk 4]
```diff
@@ -686,6 +694,26 @@ func (slamSvc *builtIn) StopSLAMProcess() error {
 	return nil
 }
 
+func (slamSvc *builtIn) getLazyPngImage(ctx context.Context, cam camera.Camera) ([]byte, error) {
```
Comment (tessavitabile) [inline services/slam/builtin/builtin.go:697]: [nit] Go style guide says this should be `getLazyPNGImage`.

File: services/slam/builtin/builtin.go (hunk 1 above)
Comment (tessavitabile) [reply services/slam/builtin/builtin.go:890]: I would suggest explicitly making a case for `Dim2d` (for Cartographer) and erroring if any other slam mode is passed in.

File: services/slam/builtin/builtin.go (hunk 3 above)
Comment (kkufieta) [reply services/slam/builtin/builtin.go:705]: Do you want the `getSimultaneousColorAndDepth` to also return `release` and call `defer release()` after the call of 

```go
images, err := slamSvc.getSimultaneousColorAndDepth(ctx, cams)
```
?

Comment (tessavitabile) [reply services/slam/builtin/builtin.go:705]: Good point, I wonder if that is a bug currently! I'm honestly not sure what the release function does, and I couldn't figure it out by reading the code. Maybe we could ask Bijan here? Or ask Eric next week.

Comment (kkufieta) [reply services/slam/builtin/builtin.go:705]: Based on our discussion (i.e. rumors that `release` could be ignored and doesn't need to be called): Removing all calls of `release` for now. Will ask Eric next week

File: services/slam/builtin/builtin.go (hunk 1 above)
Comment (JohnN193) [reply services/slam/builtin/builtin.go]: yeah that makes sense, and we can add more cases when we support them

File: services/slam/builtin/builtin.go [hunk 5]
```diff
@@ -710,6 +717,25 @@ func (slamSvc *builtIn) StopSLAMProcess() error {
...
+	lazyImg, ok := img.(*rimage.LazyEncodedImage)
+	if !ok {
+		return nil, errors.Errorf("expected lazily encoded image, got %T", lazyImg)
+	}
+	if lazyImg.MIMEType() != utils.MimeTypePNG {
+		return nil, errors.Errorf("expected mime type %v, got %v", utils.MimeTypePNG, lazyImg.MIMEType())
+	}
+	return lazyImg.RawData(), nil
+}
+
```
Comment (kkufieta) [inline services/slam/builtin/builtin.go:738]: release() needs to be called after being done using the image

File: services/slam/builtin/builtin_test.go [hunk 6]
```diff
@@ -772,10 +781,10 @@ func TestORBSLAMDataProcess(t *testing.T) {
 
 		n := 5
 		// Note: timePadding is required to allow the sub processes to be fully completed during test
-		time.Sleep(time.Millisecond * time.Duration((n)*(validDataRateMS+timePadding)))
+		time.Sleep(time.Millisecond * time.Duration((n)*(validDataRateMS+timePadding*2)))
```
Comment (tessavitabile) [inline services/slam/builtin/builtin_test.go:784]: I'm surprised that 40ms of padding is needed here, since returning the image from the injected camera and writing it to a file should be fast. Also it's possible that we'll get 6 images here, since the sleep now lasts for 5*(200+40)=1200ms. What error do you get if you only use 20ms of padding? And if you need to have 40ms of padding, do you want to change the check at the end of the test to require the number of images is greater than or equal to 5, just in case we get 6 images?

File: services/slam/builtin/builtin_test.go [hunk 7]
```diff
@@ -717,7 +717,8 @@ func TestCartographerDataProcess(t *testing.T) {
 		time.Sleep(time.Millisecond * time.Duration(validDataRateMS*2))
 		cancelFunc()
 
-		latestLoggedEntry := obs.All()[len(obs.All())-1]
+		allObs := obs.All()
+		latestLoggedEntry := allObs[len(allObs)-1]
```
Comment (tessavitabile) [inline services/slam/builtin/builtin_test.go:721]: [nit] Do you want to revert this change?

File: services/slam/builtin/builtin_test.go [hunk 8]
```diff
@@ -799,7 +808,8 @@ func TestORBSLAMDataProcess(t *testing.T) {
 		time.Sleep(time.Millisecond * time.Duration(validDataRateMS*2))
 		cancelFunc()
 
-		latestLoggedEntry := obs.All()[len(obs.All())-1]
+		obsAll := obs.All()
+		latestLoggedEntry := obsAll[len(obsAll)-1]
```
Comment (tessavitabile) [inline services/slam/builtin/builtin_test.go:812]: [nit] revert this change?

**Review by tessavitabile — COMMENTED** [review]
Nice job getting this working! A few comments/questions.

File: services/slam/builtin/builtin.go (hunk 5 above)
Comment (kkufieta) [reply services/slam/builtin/builtin.go:742]: I created a separate ticket for this, since currently `release` doesn't do anything yet, so there's some time to change the code to implement this.

File: services/slam/builtin/builtin_test.go (hunk 6 above)
Comment (kkufieta) [reply services/slam/builtin/builtin_test.go:784]: I get an error that only 4 images were saved - while 5 images are expected. Setting it to 40 ms allows it to save 5 images. Setting it to 38 would still only return 4 images. Thus, I think the risk of it returning more than 5 images is small, but requiring `>= 5` images makes sense.

File: services/slam/builtin/builtin_test.go (hunk 7 above)
Comment (kkufieta) [reply services/slam/builtin/builtin_test.go:721]: The original code caused a failed flaky test - probably because between calling `obs.All()` twice, one more log was created, and thus it threw an index out of bounds error. This fixed the problem.

**Review by tessavitabile — APPROVED** [review]
LGTM
